File: rxhttp/src/main/java/com/senierr/http/builder/UrlBuilder.java
package com.senierr.http.builder;

import java.util.LinkedHashMap;

import io.reactivex.annotations.NonNull;

/**
 * Http请求行构建器
 *
 * @author zhouchunjie
 * @date 2018/8/29
 */
public final class UrlBuilder {

    private @NonNull String url;
    private @NonNull LinkedHashMap<String, String> urlParams = new LinkedHashMap<>();

    public UrlBuilder(String url) {
        this.url = url;
    }

    public void setUrl(@NonNull String url) {
        this.url = url;
    }

    public @NonNull String getUrl() {
        return url;
    }

    public void addUrlParam(@NonNull String key, @NonNull String value) {
        urlParams.put(key, value);
    }

    public void addUrlParams(@NonNull LinkedHashMap<String, String> params) {
        urlParams.putAll(params);
    }

    public @NonNull LinkedHashMap<String, String> getUrlParams() {
        return urlParams;
    }

    public @NonNull String build() {
        if (!urlParams.isEmpty()) {
            StringBuilder strParams = new StringBuilder();
            if (url.contains("?")) {
                strParams.append("&");
            } else {
                strParams.append("?");
            }

            for (String key : urlParams.keySet()) {
                strParams.append("&").append(key).append("=").append(urlParams.get(key));
            }

            strParams.deleteCharAt(1);
            if (url.indexOf("?") == url.length() - 1) {
                strParams.deleteCharAt(0);
            }

            strParams.insert(0, url);
            url = strParams.toString();
        }
        return url;
    }
}

File: rxhttp/src/main/java/com/senierr/http/builder/UrlBuilder.kt
package com.senierr.http.builder

import android.text.TextUtils

/**
 * Http请求行构建器
 *
 * @author zhouchunjie
 * @date 2018/8/29
 */
class UrlBuilder(private val url: String) : Builder<String> {

    private var baseUrl: String? = null
    private val baseUrlParams = LinkedHashMap<String, String>()
    private var ignoreBaseUrl = false
    private var ignoreBaseUrlParams = false

    private val urlParams: LinkedHashMap<String, String> = LinkedHashMap()

    fun baseUrl(baseUrl: String?) {
        this.baseUrl = baseUrl
    }

    fun addBaseUrlParam(key: String, value: String) {
        baseUrlParams[key] = value
    }

    fun addBaseUrlParams(params: LinkedHashMap<String, String>) {
        baseUrlParams.putAll(params)
    }

    fun ignoreBaseUrl() {
        ignoreBaseUrl = true
    }

    fun ignoreBaseUrlParams() {
        ignoreBaseUrlParams = true
    }

    fun addUrlParam(key: String, value: String) {
        urlParams[key] = value
    }

    fun addUrlParams(params: LinkedHashMap<String, String>) {
        urlParams.putAll(params)
    }

    override fun build(): String {
        // 若设置了基础请求地址，且未忽略，拼接基础请求地址
        var actualUrl = url
        if (!TextUtils.isEmpty(baseUrl) && !ignoreBaseUrl) {
            actualUrl = baseUrl + url
        }

        // 若设置了基础参数，且未忽略，添加基础参数（注意添加顺序）
        val actualUrlParams = LinkedHashMap<String, String>()
        if (baseUrlParams.isNotEmpty() && !ignoreBaseUrlParams) {
            actualUrlParams.putAll(baseUrlParams)
        }
        actualUrlParams.putAll(urlParams)

        // 生成URL
        if (actualUrlParams.isNotEmpty()) {
            val strParams = StringBuilder()
            if (actualUrl.contains("?")) {
                if (actualUrl.indexOf("?") != actualUrl.lastIndex) {
                    strParams.append("&")
                }
            } else {
                strParams.append("?")
            }

            actualUrlParams.keys.forEachIndexed { index, key ->
                if (index > 0) {
                    strParams.append("&")
                }
                strParams.append(key).append("=").append(actualUrlParams[key])
            }

            strParams.insert(0, actualUrl)
            actualUrl = strParams.toString()
        }
        return actualUrl
    }
}